### PR TITLE
Add in mean for QualiMap

### PR DESCRIPTION
**Pull request by apeltzer**
If this PR is _not_ a new module
 - [x] This comment contains a description of changes (with reason)
 - [x] `CHANGELOG.md` has been updated
 - [ ] (optional but recommended): https://github.com/ewels/MultiQC_TestData contains test data for this change

### CODE REVIEW

**Comment by apeltzer** [comment]
Hm, it complains about mosdepth module broken 🤔 ...

**Comment by ewels** [comment]
Strange.. It also complains about that the Qualimap module is broken as well though:

```python 
Module qualimap raised an exception: Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/multiqc-1.8.dev0-py3.7.egg/multiqc/multiqc.py", line 546, in run
    output = mod()
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/multiqc-1.8.dev0-py3.7.egg/multiqc/modules/qualimap/qualimap.py", line 30, in __init__
    from . import QM_BamQC
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/multiqc-1.8.dev0-py3.7.egg/multiqc/modules/qualimap/QM_BamQC.py", line 154
   self.general_stats_data[s_name]['mean_coverage' = mean_coverage
                                                    ^
SyntaxError: invalid syntax
```

File: multiqc/modules/qualimap/QM_BamQC.py [hunk 1]
```diff
@@ -138,15 +138,20 @@ def parse_coverage(self, f):
...
     median_coverage = None
+    mean_coverage = None
     for thiscov, thiscount in d.items():
         cum_counts += thiscount
+        total_cov += thiscov*thiscount
         if cum_counts >= num_counts/2:
             median_coverage = thiscov
             break
+    mean_coverage = total_cov / num_counts
+    self.general_stats_data[s_name]['mean_coverage' = mean_coverage
```
Comment (ewels) [inline multiqc/modules/qualimap/QM_BamQC.py]: Missing a square bracket here:

```suggestion
    self.general_stats_data[s_name]['mean_coverage'] = mean_coverage
```

File: multiqc/modules/qualimap/QM_BamQC.py [hunk 2]
```diff
@@ -138,15 +138,20 @@ def parse_coverage(self, f):
         log.debug("Couldn't parse contents of coverage histogram file {}".format(f['fn']))
         return None
 
-    # Find median without importing anything to do it for us
+    # Find median and mean without importing anything to do it for us
     num_counts = sum(d.values())
     cum_counts = 0
+    total_cov = 0
     median_coverage = None
+    mean_coverage = None
```
Comment (ewels) [inline multiqc/modules/qualimap/QM_BamQC.py]: I don't think you need this variable? You can just assign directly to `self.general_stats_data[s_name]['mean_coverage']` at the end and save two lines of code..

File: multiqc/modules/qualimap/QM_BamQC.py [hunk 3]
```diff
@@ -138,15 +138,20 @@ def parse_coverage(self, f):
...
     for thiscov, thiscount in d.items():
         cum_counts += thiscount
+        total_cov += thiscov*thiscount
         if cum_counts >= num_counts/2:
             median_coverage = thiscov
             break
+    mean_coverage = total_cov / num_counts
```
Comment (ewels) [inline multiqc/modules/qualimap/QM_BamQC.py]: You need to handle the case of `num_counts` being zero:

```python
try:
    self.general_stats_data[s_name]['mean_coverage'] = total_cov / num_counts
except ZeroDivisionError:
    self.general_stats_data[s_name]['mean_coverage'] = 0
```

(someone, somewhere will manage this edge case, I promise..)

**Comment by ewels** [comment]
Had a look at why the mosdepth module was failing - turns out that it imports the Qualimap module and uses the same help texts for its graphs 😅 

https://github.com/ewels/MultiQC/blob/bb2cdb075a7e1588d15e57b983f12418b390fbdf/multiqc/modules/mosdepth/mosdepth.py#L12

So because your Qualimap code is broken, the mosdepth module also breaks when it tries to import it.

**Comment by apeltzer** [comment]
All added as per your request. Mean is added to general stats table, mosdepth is happy again ;-) 

BAMQC and RNAQC Modules both write the respectively parsed data now to separate output files, all named as:

`multiqc_(bamqc,rnaseq)_...`

**Comment by ewels** [comment]
Nice! Speedy work 😉 

Note that `bamqc` is the name of a [different tool](https://github.com/s-andrews/BamQC), so can the files please be `multiqc_qualimap_(bamqc,rnaseq)_...`

File: multiqc/modules/qualimap/QM_BamQC.py [hunk 4]
```diff
@@ -22,25 +22,31 @@ def parse_reports(self):
...
         parse_genome_results(self, f)
     self.qualimap_bamqc_genome_results = self.ignore_samples(self.qualimap_bamqc_genome_results)
+    self.write_data_file(self.qualimap_bamqc_genome_results, 'multiqc_bamqc_genome_results')
 
     # Coverage - coverage_histogram.txt
     self.qualimap_bamqc_coverage_hist = dict()
     for f in self.find_log_files('qualimap/bamqc/coverage', filehandles=True):
         parse_coverage(self, f)
     self.qualimap_bamqc_coverage_hist = self.ignore_samples(self.qualimap_bamqc_coverage_hist)
+    self.write_data_file(self.qualimap_bamqc_coverage_hist, 'multiqc_bamqc_coverage_hist')
```
Comment (ewels) [inline multiqc/modules/qualimap/QM_BamQC.py]: Ah sorry, I wasn't clear. No need to save data from plots - this is done already from the plot code (if requested on the command line). The only thing this needs to be added for is metrics data. Especially any that doesn't make it in to the report.

Comment (apeltzer) [reply multiqc/modules/qualimap/QM_BamQC.py]: Sorry I'm a bit lost here - which parts are missing then?

Comment (ewels) [reply multiqc/modules/qualimap/QM_BamQC.py]: Nothing missing - you added too much! If you can _remove_ the lines where you export the plot data, that would be fab 👍 (anything with `_hist`)

File: multiqc/modules/qualimap/QM_BamQC.py [hunk 5]
```diff
@@ -22,25 +22,31 @@ def parse_reports(self):
...
         parse_genome_results(self, f)
     self.qualimap_bamqc_genome_results = self.ignore_samples(self.qualimap_bamqc_genome_results)
+    self.write_data_file(self.qualimap_bamqc_genome_results, 'multiqc_qualimap_bamqc_genome_results')
 
     # Coverage - coverage_histogram.txt
     self.qualimap_bamqc_coverage_hist = dict()
     for f in self.find_log_files('qualimap/bamqc/coverage', filehandles=True):
         parse_coverage(self, f)
     self.qualimap_bamqc_coverage_hist = self.ignore_samples(self.qualimap_bamqc_coverage_hist)
+    self.write_data_file(self.qualimap_bamqc_coverage_hist, 'multiqc_qualimap_bamqc_coverage_hist')
```
Comment (ewels) [inline multiqc/modules/qualimap/QM_BamQC.py]: Please remove

File: multiqc/modules/qualimap/QM_BamQC.py [hunk 6]
```diff
@@ -22,25 +22,31 @@ def parse_reports(self):
...
+
 
     # Insert size - insert_size_histogram.txt
     self.qualimap_bamqc_insert_size_hist = dict()
     for f in self.find_log_files('qualimap/bamqc/insert_size', filehandles=True):
         parse_insert_size(self, f)
     self.qualimap_bamqc_insert_size_hist = self.ignore_samples(self.qualimap_bamqc_insert_size_hist)
+    self.write_data_file(self.qualimap_bamqc_insert_size_hist, 'multiqc_qualimap_bamqc_insert_size_hist')
```
Comment (ewels) [inline multiqc/modules/qualimap/QM_BamQC.py]: Please remove

File: multiqc/modules/qualimap/QM_BamQC.py [hunk 7]
```diff
@@ -22,25 +22,31 @@ def parse_reports(self):
...
+
 
     # GC distribution - mapped_reads_gc-content_distribution.txt
     self.qualimap_bamqc_gc_content_dist = dict()
     self.qualimap_bamqc_gc_by_species = dict()  # {'HUMAN': data_dict, 'MOUSE': data_dict}
     for f in self.find_log_files('qualimap/bamqc/gc_dist', filehandles=True):
         parse_gc_dist(self, f)
     self.qualimap_bamqc_gc_by_species = self.ignore_samples(self.qualimap_bamqc_gc_by_species)
+    self.write_data_file(self.qualimap_bamqc_gc_by_species, 'multiqc_qualimap_bamqc_gc_by_species')
```
Comment (ewels) [inline multiqc/modules/qualimap/QM_BamQC.py]: Please remove

File: multiqc/modules/qualimap/QM_RNASeq.py [hunk 8]
```diff
@@ -102,6 +102,9 @@ def parse_reports(self):
     self.qualimap_rnaseq_genome_results = self.ignore_samples(self.qualimap_rnaseq_genome_results)
     self.qualimap_rnaseq_cov_hist = self.ignore_samples(self.qualimap_rnaseq_cov_hist)
 
+    # Write data to file 
+    self.write_data_file(self.qualimap_rnaseq_genome_results, 'multiqc_qualimap_rnaseq_genome_results')
+    self.write_data_file(self.qualimap_rnaseq_cov_hist, 'multiqc_qualimap_rnaseq_cov_hist')
```
Comment (ewels) [inline multiqc/modules/qualimap/QM_RNASeq.py]: Please remove

**Comment by apeltzer** [comment]
Done, thanks for the review @ewels :-)

File: multiqc/modules/qualimap/QM_BamQC.py (hunk 4 above)
Comment (apeltzer) [reply multiqc/modules/qualimap/QM_BamQC.py]: Done :-)

**Pull request merged**